### DISCUSSION TITLE
performance is low even I added indexes

### DISCUSSION THREAD

**Discussion by malalecherocks** [supabase/supabase#9131] [General]
I was so enthusiastic with supabase, the possibility to query the DB from the client directly made me see how fast was loading everything.

My problem came when I added +10k rows to the table contact, then each query is more than 1,5seconds in the network, while it used to be 200ms.
<img width="665" alt="Captura de pantalla 2022-09-23 a las 12 49 11" src="https://user-images.githubusercontent.com/33634311/191945155-ecf79f84-6558-4508-912d-ec2074b5841f.png">

The query is:
<img width="291" alt="Captura de pantalla 2022-09-23 a las 12 43 55" src="https://user-images.githubusercontent.com/33634311/191944397-0f18fdf4-b050-42b6-afb8-4beb86caf346.png">

I also added an index:
account_id,search,id

When I query via the sql editor (Tableplus)
`select * from contact where (account_id='54511325-f1fb-4492-8e8a-b6e1cbe902f8') and (search ilike '%652225012%') order by id ASC`
It takes only 100ms, so seems to me that it's well indexed, and I'm only bringing 25 contacts via supabase client, 

even if I remove the ilike
<img width="303" alt="Captura de pantalla 2022-09-23 a las 12 56 32" src="https://user-images.githubusercontent.com/33634311/191946364-16bb5c2e-6df7-40ef-bfb6-327c4a462eee.png">

, still takes almost 2 seconds (network) to execute. And I was planning to have in this table +500k contacts. If I reduce to 1000 contacts, it goes down to 200ms.

Any suggestions?

**Comment by GaryAustin1** [comment]
Do you have RLS on this table?  What is your select policy if so?

**Reply by malalecherocks** [reply]
as usual, @GaryAustin1 has always the right answer... I just disabled the RLS of the table contact which was:

`(is_member_of_account(uid(), account_id) OR is_owner_of_account(uid(), account_id))`

and it goes blazing fast now... the problem is that I was relying a lot in RLS in my app, I'm so hype that I always want to use the latest tech possible, and I loved the idea of loading stuff directly in the client... my problem came when I started to develop my web app to have the possibility of "teams" then RLS works, but I have to use those SQL functions that it's making the database go slower (1,7s per query in contacts).

So I guess I should leave my dream of using RLS to build the app and instead call my internal API that checks permissions manually and if OK then do the query normal. It's two queries... but two 200ms queries are faster than 1,7seconds.

**Comment by GaryAustin1** [comment]
You can probably fix this by adjusting RLS policy so it does not need to run the functions on each row.  Basically you have a function return a list of accounts the user has access to. You can declare it stable and better is just SQL instead of plpgsql. 
Then you just do an IN on that result in the policy.

**Reply by malalecherocks** [reply]
you mean first create a function that returns all the teams of that account, and use that function in the RLS using the clause IN, right?

**Comment by GaryAustin1** [comment]
I’m on my phone right now with just email. Why don’t you show your function and main table columns. I’ll look when I get back. 
But the goal is you don’t want the function changing on each row.

**Reply by malalecherocks** [reply]
Here you have it... but the bigger the project it is... the more I'm realizing that RLS is made to have simple user-table rules, rather than poclicies of teams in the middle, because not only makes things more difficult to understand, but in this case I'm facing performance issues, that it might be easy to resolve with what you mentioned, but I'm aware that in the future this will make the project more difficult to maintain...


Here you have it...

<img width="660" alt="Captura de pantalla 2022-09-23 a las 22 10 01" src="https://user-images.githubusercontent.com/33634311/192049665-e1874fea-7b8b-4b50-8c3a-3a30db7f8011.png">
<img width="664" alt="Captura de pantalla 2022-09-23 a las 22 10 14" src="https://user-images.githubusercontent.com/33634311/192049676-b196c355-5077-4fb9-85e6-3d414c5b68c5.png">
<img width="662" alt="Captura de pantalla 2022-09-23 a las 22 10 30" src="https://user-images.githubusercontent.com/33634311/192049681-0aee00df-00f8-428b-95bc-c8f577b799e7.png">
<img width="1142" alt="Captura de pantalla 2022-09-23 a las 22 11 05" src="https://user-images.githubusercontent.com/33634311/192049685-c20df372-2ecc-46d5-8754-2f64ffc98dc9.png">
<img width="656" alt="Captura de pantalla 2022-09-23 a las 22 12 04" src="https://user-images.githubusercontent.com/33634311/192049688-7cac07ad-059d-4249-ad78-93baf418bc5d.png">
<img width="655" alt="Captura de pantalla 2022-09-23 a las 22 12 14" src="https://user-images.githubusercontent.com/33634311/192049690-cb340d15-4bfd-4cec-83c5-02a678bf4c66.png">

**Reply by steve-chavez** [reply]
Can you run and share the result of this query:

```sql
select proretset, provolatile from pg_catalog.pg_proc where proname in ('is_owner_of_account', 'is_member_of_account');
```

This will tell us if the functions are [inlinable](https://wiki.postgresql.org/wiki/Inlining_of_SQL_functions).

**Reply by malalecherocks** [reply]
This is the response, no clue what you are trying to do xD

<img width="248" alt="Captura de pantalla 2022-09-24 a las 9 49 48" src="https://user-images.githubusercontent.com/33634311/192086950-aa1f46f5-6efe-4118-9eb7-92d4a574fe9e.png">

**Reply by steve-chavez** [reply]
With that basically we can check if the functions you created will run for every row(as Gary mentioned) - which is true from the results you've shared.

To avoid that the function needs to be created as stable:

```sql
create or replace function is_member_of_account() 
returns .. STABLE as $$
$$;
```

**Comment by GaryAustin1** [comment]
For this particular RLS policy do you even need the owner check?  Is the owner not also a member?

I'm not as versed as Steve is on this stuff, but I'm not sure how much stable helps when it appears you are passing in the row's account_id to the function, and it would change on each row.   This is why I suggested looking at something like `account_id in my_accounts()`.  Where the my_accounts function uses uid(), can be declared stable, and returns a list of accounts the user is a member of.  BUT without knowing your data this could cause issues if a user is a member of "a lot" of accounts... where "a lot" is unknown to me, but 10's is certainly fine...

Edit:  in this case it looks like you have an account_id filter so the stable on your current functions might be fine as you are not changing the account input.
